File: Python/Medium/1d-programming/word_break.py
class Solution:
    def wordBreak(s, word_dict):
        n=len(s)
        dp = [False]*(n+1)
        dp[0] = True #Empty string is always segmented

        for i in range(1, n+1):
            for word in word_dict:
                print(dp[i-len(word)])
                if i-len(word)> 0 and dp[i-len(word)] and s[i-len(word):i] == word:
                    dp[i]= True
                    break

        return dp[n]



s = "leetcode"
wordDict = ['leet', 'code']

n=6
dp = [False]*(n+1)
if dp[1-6]:
    print('its a neg')

print(Solution.wordBreak(s, wordDict))

File: JavaScript/Medium/1d_programming/word_break.js
var wordBreak = function (s, word_dict){
    let n=s.length
    let dp = new Array(n+1).fill(false)
    dp[0] = true

    for (let i = 1; i<n+1; i++) {
        for (word of word_dict) {
            if (i-word.length >= 0 && dp[i-word.length] && s.slice(i-word.length, i) == word) {
                dp[i]= true
                break
            }
        }
    }

    return dp[n]
}